### PR TITLE
Add analyze-project composite action to validate a Python project

### PR DESCRIPTION
**Pull request by dixonjoel**
### What does this Pull Request accomplish?

Add a analyze-project action that can be re-used in other repos.

### Why should this Pull Request be merged?

Reuse the .yml files instead of copying them around to every repo.

### What testing has been done?

Tested by calling the new action from this branch in datastore-python and nipanel-python repos

Checks from using this action in nipanel:
https://github.com/ni/nipanel-python/actions/runs/17921957525

Checks from using this action in datastore-python:
https://github.com/ni/datastore-python/actions/runs/17921449681

### CODE REVIEW

File: workflows/analyze-package.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+          key: ${{ inputs.package-name }}-${{ runner.os }}-py${{ steps.setup-python.outputs.python-version }}-${{ hashFiles(format('{0}/poetry.lock', inputs.project-directory)) }}
+      - name: Install ${{ steps.get_package_info.outputs.name }}
+        run: |
+          if [ "${{ inputs.install-extras }}" != "" ]; then
+            poetry install -v --extras "${{ inputs.install-extras }}"
+          else
+            poetry install -v
+          fi
+      - name: Lint
+        run: poetry run ni-python-styleguide lint
```
Comment (mshafer-NI) [inline workflows/analyze-package.yml:61]: in PR's/CI's, it may be useful to either default to `-vv`

File: workflows/analyze-package.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      - name: Mypy static analysis (Linux)
+        run: poetry run mypy
+      - name: Mypy static analysis (Windows)
+        run: poetry run mypy --platform win32
```
Comment (mshafer-NI) [inline workflows/analyze-package.yml:65]: I'm not sure about this...

If people are already running tests on Linux and on Windows, then us explicitly doing both is overhead; however, if user's don't matrix, then this is usefull...

maybe make line 62 (without a name) the default and line 64 (with name and the platform passed to 65) based on a "mypyAlsoCheck" optional parameter?

File: workflows/analyze-package.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      - name: Bandit security checks
+        run: poetry run bandit -c pyproject.toml -r src/
```
Comment (mshafer-NI) [inline workflows/analyze-package.yml:67]: Let's not add this here. Once https://github.com/ni/python-styleguide/pull/204 has the right sign-off's (which may be soon), running ni-python-styleguide will cover this.

File: workflows/analyze-package.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        with:
+          python-platform: Linux
+          version: PATH
+          working-directory: ${{ inputs.project-directory }}
+      - name: Pyright static analysis (Windows)
+        uses: jakebailey/pyright-action@b5d50e5cde6547546a5c4ac92e416a8c2c1a1dfe # v2.3.2
+        with:
+          python-platform: Windows
+          version: PATH
+          working-directory: ${{ inputs.project-directory }}
```
Comment (mshafer-NI) [inline workflows/analyze-package.yml:81]: Same comment as mypy (on platforms)...

File: workflows/analyze-package.yml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        run: |
+          if [ "${{ inputs.install-extras }}" != "" ]; then
+            poetry install -v --extras "${{ inputs.install-extras }}"
+          else
+            poetry install -v
+          fi
+      - name: Lint
+        run: poetry run ni-python-styleguide lint
+      - name: Mypy static analysis (Linux)
+        run: poetry run mypy
```
Comment (mshafer-NI) [inline workflows/analyze-package.yml:63]: Currently, our policy is that mypy is optional. Maybe this should follow what our internal pipeline does and check if the pyproject.toml lists mypy as a dep?

File: workflows/analyze-package.yml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      - name: Mypy static analysis (Windows)
+        run: poetry run mypy --platform win32
+      - name: Bandit security checks
+        run: poetry run bandit -c pyproject.toml -r src/
+      - name: Add virtualenv to the path for pyright-action
+        run: echo "$(poetry env info --path)/bin" >> $GITHUB_PATH
+      - name: Pyright static analysis (Linux)
```
Comment (mshafer-NI) [inline workflows/analyze-package.yml:70]: Pyright is technically "new", anything that is "new" should definitely go through a phase of being optional. Maybe this should follow what our internal pipeline does and check if the pyproject.toml lists pyright as a dep?

File: workflows/analyze-package.yml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        default: ${{ github.workspace }}
+      install-extras:
+        description: 'List of Poetry extras to install (comma separated)'
+        default: ''
+        required: false
+        type: string
+
+jobs:
+  check_analyzers:
+    name: Check analyzers for ${{ inputs.package-name }}
```
[the remark this answers is not in the document]
Comment (mshafer-NI) [reply workflows/analyze-package.yml:22]: So..., not everything we're running is a "static analysis" tool...

Yes, we have some, we also have formatters, (I would add linters and type checkers to this, but these could be argued to be static analysis or possibly import analysis).

Maybe..., "validate" or "check quality" or "run checks" or "check code quality" or simply "run checks" ...,

File: workflows/analyze-package.yml (hunk 5 above)
Comment (mshafer-NI) [reply workflows/analyze-package.yml:63]: Because these are internal sources, I sent you some links offline

File: workflows/analyze-package.yml (hunk 2 above)
Comment (bkeryan) [reply workflows/analyze-package.yml:65]: This is based on our "check_nifoo.yml" workflows, which only run on Linux.

Perhaps we should run it in a matrix. That would eliminate some mypy/pyright false positives. However, Joel is trying to refactor the workflows we already have.

Comment (bkeryan) [reply workflows/analyze-package.yml:65]: Either way, it could make sense to have an input for the supported platforms.

Or convert this from a reusable workflow to a composite action, so the caller can control where it runs.

File: workflows/analyze-package.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply workflows/analyze-package.yml:81]: Unit tests need a large matrix of Python versions. Nps/mypy/pyright do not.

[the remark this answers is not in the document]
Comment (mshafer-NI) [reply workflows/analyze-package.yml:81]: although..., we've learned that mypy does. We pass "--python-version {x.y}" (overwriting what is in pyproj.toml) to check for typing / syntax issues with changes in stdlib / etc.

[the remark this answers is not in the document]
Comment (bkeryan) [reply workflows/analyze-package.yml:81]: That's fair, I have been ignoring that aspect.

Also, I have been ignoring oldest compatible version testing for mypy, because I have no expectation of mypy passing with anything other than the very latest versions of numpy, types-*, etc.

File: workflows/analyze-package.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply workflows/analyze-package.yml:22]: > We'll need an input for `package-name`

I previously suggested passing `project-path` and asking Poetry for the `package-name`.

File: workflows/check_analyzers.yml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            poetry install -v
+          fi
+      - name: Lint
+        run: poetry run ni-python-styleguide lint
+      - name: Mypy static analysis (Linux)
+        run:  poetry run mypy
+      - name: Mypy static analysis (Windows)
+        run:  poetry run mypy --platform win32
+      - name: Bandit security checks
+        run:  poetry run bandit -c pyproject.toml -r src/
```
[the remark this answers is not in the document]
Comment (dixonjoel) [reply workflows/check_analyzers.yml]: For this PR, I'll leave it out and just use 'poetry run bandit -c pyproject.toml -r'. I guess I'll document that requirement with a comment somewhere in this file.

[the remark this answers is not in the document]
Comment (dixonjoel) [reply workflows/check_analyzers.yml]: Took out the 'src/' and added this comment `# Assumes there are 'targets' in pyproject.toml for Bandit`

File: workflows/check_analyzers.yml [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        run: poetry check --lock
+      - name: Cache virtualenv
+        uses: actions/cache@0400d5f644dc74513175e3cd8d07132dd4860809 # v4.2.4
+        with:
+          path: ${{ inputs.package-base-path }}/${{ inputs.package-name }}/.venv
+          key: ${{ inputs.package-name }}-${{ runner.os }}-py${{ steps.setup-python.outputs.python-version }}-${{ hashFiles(format('{0}/{1}/poetry.lock', inputs.package-base-path, inputs.package-name)) }}
+      - name: Install ${{ inputs.package-name }}
+        run: |
+          if [ "${{ inputs.install-drivers }}" = "true" ]; then
+            poetry install -v --extras drivers
```
[the remark this answers is not in the document]
Comment (dixonjoel) [reply workflows/check_analyzers.yml]: Updated to 'extras' and changed the logic to add `-E "extra1" -E "extra2"` etc

File: workflows/check_analyzers.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply workflows/check_analyzers.yml]: Sorry, ni-python-styleguide is going to do it the other way (assume .)

File: workflows/analyze-package.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply workflows/analyze-package.yml]: This may also involve letting the caller do the setup-python and setup-poetry so they can specify the versions.

[the remark this answers is not in the document]
Comment (bkeryan) [reply workflows/analyze-package.yml]: I also think we ought to try out the matrixed checks in some other components like nitypes/nipanel/ni-apis-python before we bake it in to the single-sourced version

File: workflows/check_analyzers.yml [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        id: setup-python
+      - name: Set up Poetry
+        uses: ni/python-actions/setup-poetry@f0276f7f58868ec0d0d1a86377287c9e6fe0c6e7 # v0.5.0
+      - name: Check for lock changes
+        run: poetry check --lock
+      - name: Cache virtualenv
+        uses: actions/cache@0400d5f644dc74513175e3cd8d07132dd4860809 # v4.2.4
+        with:
+          path: ${{ inputs.package-base-path }}/${{ inputs.package-name }}/.venv
+          key: ${{ inputs.package-name }}-${{ runner.os }}-py${{ steps.setup-python.outputs.python-version }}-${{ hashFiles(format('{0}/{1}/poetry.lock', inputs.package-base-path, inputs.package-name)) }}
```
[the remark this answers is not in the document]
Comment (dixonjoel) [reply workflows/check_analyzers.yml]: Is that something we want to do as a prerequisite to this new action? Or after the fact?

File: workflows/analyze-package.yml (hunk 5 above)
Comment (dixonjoel) [reply workflows/analyze-package.yml]: One simple thing Copilot mentioned is just to check if "mypy" appears in the pyproject.toml file. This means we wouldn't need all the logic in the links you sent for getting the dependencies dynamically. What do you think about that idea?
```
- name: Check for mypy dependency
  id: has_mypy
  run: |
    if grep -E 'mypy' "${{ inputs.project-directory }}/pyproject.toml"; then
      echo "found=true" >> $GITHUB_OUTPUT
    else
      echo "found=false" >> $GITHUB_OUTPUT
    fi
```

File: workflows/check_analyzers.yml
[the remark this answers is not in the document]
Comment (dixonjoel) [reply workflows/check_analyzers.yml:1]: Switching over to a composite action.

File: workflows/analyze-package.yml (hunk 2 above)
Comment (dixonjoel) [reply workflows/analyze-package.yml]: Switching over to a composite action.

File: .github/workflows/test_actions.yml [hunk 11]
```diff
@@ -212,6 +212,63 @@ jobs:
...
+        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+      - name: Set up Python
+        uses: ./setup-python
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Set up Poetry
+        uses: ./setup-poetry
+      - name: Create project
+        run: poetry new test-project --python ">=3.7,<4.0"
+      - name: Add docstring to source files
```
[the remark this answers is not in the document]
Comment (bkeryan) [reply .github/workflows/test_actions.yml]: Yeah, it may be easier to check a project into source control.

[the remark this answers is not in the document]
Comment (dixonjoel) [reply .github/workflows/test_actions.yml]: The goal is to sanity check the analyze-project action. Another option Copilot suggested is to use a project template of some kind. I'm open to other ideas, but for now I'm trying to make this work.

File: workflows/check_analyzers.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (dixonjoel) [reply workflows/check_analyzers.yml]: Not tackling this in this PR.

File: workflows/analyze-package.yml (hunk 4 above)
Comment (dixonjoel) [reply workflows/analyze-package.yml]: I made a composite action and tested it from the datastore-python repo with a matrix. I think that covers the platform concerns.

[the remark this answers is not in the document]
Comment (dixonjoel) [reply workflows/analyze-package.yml]: Is a composite action now.

File: .github/workflows/test_actions.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (dixonjoel) [reply .github/workflows/test_actions.yml]: Test project is now under .github/test_projects/minimal

File: .github/workflows/test_actions.yml [hunk 12]
```diff
@@ -212,6 +212,30 @@ jobs:
...
+        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+      - name: Set up Python
+        uses: ./setup-python
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Set up Poetry
+        uses: ./setup-poetry
+      - name: Copy test project files
+        run: |
+          cp -r ./.github/test_projects/minimal ./minimal
```
Comment (bkeryan) [inline .github/workflows/test_actions.yml:233]: Why copy the project? This sounds like continued fallout from an unnecessary Copilot suggestion (create a project template). I would rather set `project-directory: ./.github/test_projects/minimal` and check it in-place.

File: analyze-project/action.yml [hunk 13]
```diff
@@ -0,0 +1,94 @@
+name: Analyze Python Project
```
Comment (bkeryan) [inline analyze-project/action.yml:1]: The other action names are in sentence case and match the action subdirectory. We don't need to Smurf-name everything with "Python".
```suggestion
name: Analyze project
```

File: analyze-project/action.yml [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+description: >
+  This workflow checks the code quality of a Python project using various
+  analyzers like linters and type checkers, including ni-python-styleguide,
+  mypy, bandit, and pyright. It is designed to be reusable across different
+  projects and can be easily integrated into existing CI/CD pipelines.
```
Comment (bkeryan) [inline analyze-project/action.yml:6]: - This action doesn't support bandit anymore.
- It would be simpler to say "This workflow analyzes the code quality..." and drop "analyzers".
- The last sentence describes GitHub Actions in general, not this specific action, and should be omitted.

File: analyze-project/action.yml [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        echo "version=$version" >> $GITHUB_OUTPUT
+      shell: bash
+    - name: Check for lock changes
+      run: poetry check --lock -C "${{ inputs.project-directory }}"
+      shell: bash
+    - name: Cache virtualenv
+      uses: actions/cache@0400d5f644dc74513175e3cd8d07132dd4860809 # v4.2.4
+      with:
+        path: ${{ inputs.project-directory }}/.venv
+        key: ${{ inputs.package-name }}-${{ runner.os }}-py${{ steps.setup-python.outputs.python-version }}-${{ hashFiles(format('{0}/poetry.lock', inputs.project-directory)) }}
```
Comment (bkeryan) [inline analyze-project/action.yml:38]: There is no inputs.package-name or steps.setup-python in this action, so those parts of the cache-key evaluate to empty strings. Use steps.get_package_info.outputs.name and env.pythonVersion instead. Please look at the cache key in the log and make sure it's not missing parts.

File: analyze-project/action.yml [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+      run: |
+        if [ "${{ inputs.extras }}" != "" ]; then
+          EXTRAS_ARGS=""
+          for extra in ${{ inputs.extras }}; do
+            EXTRAS_ARGS="$EXTRAS_ARGS -E $extra"
+          done
+          poetry install -v $EXTRAS_ARGS
+        else
+          poetry install -v
+        fi
```
Comment (bkeryan) [inline analyze-project/action.yml:49]: Poetry accepts space-delimited extras. Remember https://github.com/ni/ni-apis-python/blob/main/.github/workflows/check_analyzers.yml#L47 ?

```suggestion
        if [ -n "${{ inputs.extras }}" ]; then
          poetry install -v --extras '${{ inputs.extras }}'
        else
          poetry install -v
        fi
```

File: workflows/analyze-package.yml (hunk 5 above)
Comment (bkeryan) [reply workflows/analyze-package.yml]: For posterity: we discussed this offline on Thursday.

- The code that @mshafer-NI linked uses toml or tomlkit to parse the pyproject.toml file, which I would prefer to avoid, especially in light of the project.dependencies migration.
- Using "grep" is likely to cause false positives if the package has references to tool.mypy or dependencies that contain mypy, like mypy-protobuf (which is for generating type stubs and may be used without actually running mypy on the project).
- Using "pip show mypy" and checking the exit code would be simple, but this doesn't work.
- Using importlib.metadata is a little more verbose (you have to catch an exception) but it seems accurate and portable.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: analyze-project/action.yml [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        echo "name=$name" >> $GITHUB_OUTPUT
+        echo "version=$version" >> $GITHUB_OUTPUT
+      shell: bash
+    - name: Check for lock changes
+      run: poetry check --lock -C "${{ inputs.project-directory }}"
+      shell: bash
+    - name: Cache virtualenv
+      uses: actions/cache@0400d5f644dc74513175e3cd8d07132dd4860809 # v4.2.4
+      with:
+        path: ${{ inputs.project-directory }}/.venv
```
Comment (jfriedri-ni) [inline analyze-project/action.yml:37]: This folder location relies on clients using a `poetry.toml` file that sets

```toml
[virtualenvs]
in-project = true
```

which is the prevailing but unenforced pattern, usually propagated via copy-paste.

Do we need to validate or parameterize this?

File: analyze-project/action.yml [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        key: ${{ inputs.package-name }}-${{ runner.os }}-py${{ steps.setup-python.outputs.python-version }}-${{ hashFiles(format('{0}/poetry.lock', inputs.project-directory)) }}
+    - name: Install ${{ steps.get_package_info.outputs.name }}
+      run: |
+        if [ "${{ inputs.extras }}" != "" ]; then
+          EXTRAS_ARGS=""
+          for extra in ${{ inputs.extras }}; do
+            EXTRAS_ARGS="$EXTRAS_ARGS -E $extra"
+          done
+          poetry install -v $EXTRAS_ARGS
```
Comment (jfriedri-ni) [inline analyze-project/action.yml]: If the input for this action is already space-separated, we can detect a non-empty value and directly invoke poetry without re-rolling the parameters into a variable.

https://github.com/ni/ni-apis-python/blob/main/.github/workflows/check_analyzers.yml#L50

**Review by jfriedri-ni — APPROVED** [review]

File: analyze-project/action.yml (hunk 17 above)
Comment (bkeryan) [reply analyze-project/action.yml:37]: Yes, we should use `poetry env info --path`

File: analyze-project/action.yml (hunk 18 above)
Comment (dixonjoel) [reply analyze-project/action.yml]: This has been addressed. @bkeryan had made the same comment.

File: analyze-project/action.yml [hunk 19]
```diff
@@ -19,31 +19,29 @@ inputs:
...
       run: |
-        result=$(poetry version -C "${{ inputs.project-directory }}")
+        result=$(poetry version)
         name=$(echo "$result" | awk '{print $1}')
         version=$(echo "$result" | awk '{print $2}')
-        echo "name=$name" >> $GITHUB_OUTPUT
-        echo "version=$version" >> $GITHUB_OUTPUT
+        path=$("poetry env info --path")
+        echo "name=$name" >> "$GITHUB_OUTPUT"
+        echo "version=$version" >> "$GITHUB_OUTPUT"
```
Comment (bkeryan) [inline analyze-project/action.yml:30]: also `echo "venv-path=$venv_path" >> "$GITHUB_OUTPUT"`

File: analyze-project/action.yml [hunk 20]
```diff
@@ -19,31 +19,29 @@ inputs:
...
+    - name: Get project name and version
+      id: get_project_info
       run: |
-        result=$(poetry version -C "${{ inputs.project-directory }}")
+        result=$(poetry version)
         name=$(echo "$result" | awk '{print $1}')
         version=$(echo "$result" | awk '{print $2}')
-        echo "name=$name" >> $GITHUB_OUTPUT
-        echo "version=$version" >> $GITHUB_OUTPUT
+        path=$("poetry env info --path")
```
[the remark this answers is not in the document]
Comment (bkeryan) [reply analyze-project/action.yml:28]: BTW I think I previously meant to suggest `path="$(poetry env info --path)"` but I don't know if those quotes are necessary.

File: analyze-project/README.md [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+### `extras`
+
+If there are extras you need to install from your pyproject.toml, specify a space-separated list
+of extra groups to install. For example, 
+
+```yaml
+- uses: ni/python-actions/analyze-project@v0.2
+  with:
+    project-directory: ${{ github.workspace }}/packages/myproject
+    extras: 'docs drivers'
```
Comment (bkeryan) [inline analyze-project/README.md:44]: This action didn't exist in v0.2, so let's use the floating v0 tag. 

I can't specify a suggestion here because the triple backticks confuses Github.

File: analyze-project/action.yml [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+    # E.g. "docs drivers"
+    description: 'List of Poetry extras to install (separated by spaces)'
+    default: ''
+    required: false
+    type: string
+
+runs:
+  using: composite
+  steps:
+    - name: Get project name and version
```
Comment (bkeryan) [inline analyze-project/action.yml:22]:
```suggestion
    - name: Get project info
```

**Review by bkeryan — APPROVED** [review]
Approved with suggestions

File: .github/workflows/test_actions.yml [hunk 23]
```diff
@@ -212,6 +212,28 @@ jobs:
...
+      - name: Set up Poetry
+        uses: ./setup-poetry
+      - name: Analyze Python Project
+        uses: ./analyze-project
+        with:
+          project-directory: ${{ github.workspace }}/.github/test_projects/minimal
+
   # This job is intended to combine the test results so we don't have to list
   # each matrix combination in the required status check settings. There are a
   # lot of corner cases that make this harder than it should be; see See
```
Comment (bkeryan) [inline .github/workflows/test_actions.yml:239]: Add your new job to test_results.needs

File: .github/workflows/test_actions.yml [hunk 24]
```diff
@@ -212,6 +212,28 @@ jobs:
           project-directory: test-project
           expected-version: 1.0.2.dev1
 
+  test_analyze_project:
+    name: Test analyze-project
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
```
Comment (bkeryan) [inline .github/workflows/test_actions.yml:219]: This was for debugging, right?

```suggestion
```

File: .github/workflows/test_actions.yml [hunk 25]
```diff
@@ -11,7 +11,7 @@ jobs:
     strategy:
       matrix:
         os: [windows-latest, ubuntu-latest, macos-latest]
-        python-version: [3.9, '3.10', 3.11, 3.12, 3.13, 3.13t, pypy3.10, pypy3.11]
+        python-version: [3.9, '3.10', 3.11, 3.12, 3.13, 3.13, pypy3.10, pypy3.11]
```
[the remark this answers is not in the document]
Comment (dixonjoel) [reply .github/workflows/test_actions.yml:14]: Oh, I see. I had used Copilot to update some of these and I thought it had put that in for some reason.

[the remark this answers is not in the document]
Comment (dixonjoel) [reply .github/workflows/test_actions.yml]: Nevermind. I see I'm adding 3.13t where it wasn't before...

**Review by mshafer-NI — APPROVED** [review]
